source/slang/ir-legalize-types.cpp

```diff
@@ -567,40 +739,87 @@ static LegalVal legalizeInst(
     return legalVal;
 }
 
+static void addParamType(IRFuncType * ftype, LegalType t)
+{
+    switch (t.flavor)
+    {
+    case LegalType::Flavor::simple:
+        ftype->paramTypes.Add(t.obj.As<Type>());
+        break;
+    case LegalType::Flavor::implicitDeref:
+    {
+        auto imp = t.obj.As<ImplicitDerefType>();
+        addParamType(ftype, imp->valueType);
+        break;
+    }
+    case LegalType::Flavor::tuple:
+    {
+        auto tup = t.obj.As<TupleType>();
+        for (auto & elem : tup->elements)
+            addParamType(ftype, elem.type);
+    }
+    break;
+    default:
+        SLANG_ASSERT(false);
+    }
+}
+
 static void legalizeFunc(
     TypeLegalizationContext*    context,
     IRFunc*                     irFunc)
 {
     // Overwrite the function's type with
     // the result of legalization.
-    irFunc->type = legalizeSimpleType(context, irFunc->type);
-
+    auto newFuncType = new IRFuncType();
+    newFuncType->setSession(context->session);
+    auto oldFuncType = irFunc->type.As<IRFuncType>();
+    newFuncType->resultType = legalizeSimpleType(context, oldFuncType->resultType);
+    for (auto & paramType : oldFuncType->paramTypes)
```

Comment:
Shouldn’t we just call `legalizeType` here and have that function include special logic for function types? That is, we can have function types arise in places other than as the type of `IRFunc`s.